### PR TITLE
Add `GriptapeCloudConversationMemoryDriver`

### PR DESCRIPTION
**Pull request by vachillo**
- [x] I have read and agree to the contributing guidelines for [submitting new pull requests](https://github.com/griptape-ai/griptape?tab=readme-ov-file#submitting-pull-requests).

## Describe your changes
- Add `GriptapeCloudConversationMemoryDriver`
- Fix recursive loading issue in other conversation memory drivers

### CODE REVIEW

File: griptape/clients/griptape_cloud_api_client.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        default=None,
+        kw_only=True,
+    )
+    structure_run_id: Optional[str] = field(
+        default=Factory(lambda: os.getenv("GT_CLOUD_STRUCTURE_RUN_ID")), kw_only=True
+    )
+
+    def __attrs_post_init__(self) -> None:
+        if self.api_key is None:
+            raise ValueError("GriptapeCloudApiClient requires an API key")
```
Comment (collindutter) [inline griptape/clients/griptape_cloud_api_client.py:30]: Should be done via a validator.

File: griptape/clients/griptape_cloud_api_client.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        self.headers = dict_merge(
+            {"Authorization": f"Bearer {self.api_key}"},
+            self.headers if self.headers is not None else {},
+        )
```
Comment (collindutter) [inline griptape/clients/griptape_cloud_api_client.py:34]: Do we want to merge this in? What if they provide a different key in their `headers`? Definitely, but this should probably be renamed something like `extra_headers` if we're going to add them in.

File: griptape/clients/griptape_cloud_api_client.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        path = path.lstrip("/")
+        response = requests.post(f"{self.base_url}/{path}", headers=self.headers, json=data)
+        response.raise_for_status()
+        return response.json()
+
+    def patch(self, path: str, data: dict) -> dict:
+        path = path.lstrip("/")
+        response = requests.patch(f"{self.base_url}/{path}", headers=self.headers, json=data)
+        response.raise_for_status()
+        return response.json()
```
Comment (collindutter) [inline griptape/clients/griptape_cloud_api_client.py:52]: We should use `urllib.parse.urljoin` instead.

File: griptape/clients/griptape_cloud_api_client.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        kw_only=True,
+    )
+    api_key: Optional[str] = field(default=Factory(lambda: os.getenv("GT_CLOUD_API_KEY")), kw_only=True)
+    headers: dict = field(
+        default=None,
+        kw_only=True,
+    )
+    structure_run_id: Optional[str] = field(
+        default=Factory(lambda: os.getenv("GT_CLOUD_STRUCTURE_RUN_ID")), kw_only=True
+    )
```
Comment (collindutter) [inline griptape/clients/griptape_cloud_api_client.py:26]: Will all uses of the client use this?

File: griptape/clients/griptape_cloud_api_client.py (hunk 2 above)
Comment (vachillo) [reply griptape/clients/griptape_cloud_api_client.py:34]: i feel like its not intended to break the client if you want to add headers. i dont event really know why we provide the option to begin with because the API doesnt accept extra headers. if you want to give your own `Authorization` itll override, else itll just add it in

File: griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+from griptape.memory.structure import BaseConversationMemory, ConversationMemory, Run
+from griptape.mixins import ExponentialBackoffMixin
+
+
+@define(kw_only=True)
+class GriptapeCloudConversationMemoryDriver(BaseConversationMemoryDriver, ExponentialBackoffMixin):
+    id: str = field(
+        default=Factory(lambda self: os.getenv("GT_CLOUD_THREAD_ID", self._get_id()), takes_self=True),
+        metadata={"serializable": True},
+    )
```
Comment (collindutter) [inline griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py:21]: Should be renamed to `thread_id`.

File: griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    api_key: str = field(default=None, metadata={"serializable": False})
```
Comment (collindutter) [inline griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py:22]: Should pull from env var.

File: griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    base_url: str = field(default=None, metadata={"serializable": True})
```
Comment (collindutter) [inline griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py:23]: Should pull from env var.

File: griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    structure_run_id: str = field(default=None, metadata={"serializable": True})
```
Comment (collindutter) [inline griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py:24]: Is this necessary? Threads are their own thing.

File: griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+                api_key=self.api_key, base_url=self.base_url, structure_run_id=self.structure_run_id
+            ),
+            takes_self=True,
+        ),
+        metadata={"serializable": False},
+    )
+
+    def _get_id(self) -> Optional[str]:
+        thread_id = None
+        for attempt in self.retrying():
```
Comment (collindutter) [inline griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py:37]: None of the other drivers implement retrying, this should probably be done at the base level in a separate PR.

File: griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+            with attempt:
+                thread_id = self.client.post("/threads", {}).get("thread_id")
+
+        return thread_id
+
+    def store(self, memory: BaseConversationMemory) -> None:
+        messages = [{"input": json.dumps(run.input), "output": json.dumps(run.output)} for run in memory.runs]
+        for attempt in self.retrying():
+            with attempt:
```
Comment (collindutter) [inline griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py:46]: Same, let's not attempt in the child classes.

File: griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+                self.client.patch(f"/threads/{self.id}", {"messages": messages})
+
+    def load(self) -> Optional[BaseConversationMemory]:
+        response = None
+        for attempt in self.retrying():
+            with attempt:
```
Comment (collindutter) [inline griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py:52]: Attempt

File: griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+                response = self.client.get(f"/threads/{self.id}/messages")
+        if response is None:
+            return None
+        messages: list[dict] = response.get("messages", [])
+        messages = sorted(messages, key=lambda m: m["index"])
```
Comment (collindutter) [inline griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py:57]: We should not sort client side, the cloud should return sorted.

File: griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+        runs = [
+            Run(
+                id=m["message_id"], input=BaseArtifact.from_dict(m["input"]), output=BaseArtifact.from_dict(m["output"])
+            )
+            for m in messages
+        ]
+        return ConversationMemory(runs=runs, driver=self)
```
Comment (collindutter) [inline griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py:64]: Needs to support other Conversation Memory types like `SummaryConversationMemory`. Fields like `SummaryConversationMemory.summary` should be stored in thread metadata.

File: griptape/clients/griptape_cloud_api_client.py (hunk 2 above)
Comment (vachillo) [reply griptape/clients/griptape_cloud_api_client.py]: made headers `init=False`. theres no reason to give headers at this time.

File: griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py (hunk 12 above)
Comment (vachillo) [reply griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py]: it should, but with the inclusion of `index` this is just a double-check

File: griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+            raise RuntimeError(f"Error getting messages for thread {self.thread_id}")
+
+        messages = sorted(messages_response.get("messages", []), key=lambda m: m["index"])
+
+        # retrieve the Thread to get the metadata
+        thread_response = requests.get(self._get_url(f"/threads/{self.thread_id}"), headers=self.headers)
+        thread_response.raise_for_status()
+        thread_response = thread_response.json()
+
+        from griptape.memory.structure import Run
```
Comment (collindutter) [inline griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py:96]: Is this to avoid circular dependency? If so, move to top of function.

File: griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+
+        runs = [
+            Run(
+                id=m["message_id"],
+                input=BaseArtifact.from_dict(json.loads(m["input"])),
+                output=BaseArtifact.from_dict(json.loads(m["output"])),
+            )
+            for m in messages
+        ]
```
Comment (collindutter) [inline griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py:105]: Can be simplified to:

```python
        runs = [
            Run(
                id=m["message_id"],
                input=BaseArtifact.from_json(m["input"]),
                output=BaseArtifact.from_json(m["output"]),
            )
            for m in messages
        ]
```

File: griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+
+            return BaseConversationMemory.from_dict(
+                {
+                    **json.loads(thread_response.get("metadata")),
+                    "runs": [run.to_dict() for run in runs],
+                    "autoload": False,
+                }
+            )
+        # no metadata found, return a new ConversationMemory object
+        from griptape.memory.structure import ConversationMemory
```
Comment (collindutter) [inline griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py:122]: Move to top of function

File: griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+            )
+            for m in messages
+        ]
+        metadata = thread_response.get("metadata")
+
+        # the metadata will contain the serialized
+        # ConversationMemory object with the runs removed
+        # autoload=False to prevent recursively loading the memory
+        if metadata is not None and metadata != {}:
+            from griptape.memory.structure import BaseConversationMemory
```
Comment (collindutter) [inline griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py:112]: Move to top of function.

File: griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+
+            return BaseConversationMemory.from_dict(
+                {
+                    **json.loads(thread_response.get("metadata")),
+                    "runs": [run.to_dict() for run in runs],
```
Comment (collindutter) [inline griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py:117]: Is there any point in creating the `Run`s earlier in the function? Why not just pass in the messages here?

File: griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py [hunk 19]
```diff
@@ -0,0 +1,133 @@
...
+        messages = [
+            {"input": json.dumps(run.input.to_dict()), "output": json.dumps(run.output.to_dict())}
+            for run in memory.runs
+        ]
+
+        # serialize the metadata to a json string
+        # remove runs because they are already stored as Messages
+        metadata = memory.to_dict()
+        del metadata["runs"]
+        metadata = json.dumps(metadata)
```
Comment (collindutter) [inline griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py:70]: More of a comment on Cloud but it seems like `metadata` should be a dictionary. What do we gain from it being a string?

File: griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py [hunk 20]
```diff
@@ -0,0 +1,133 @@
...
+        # ConversationMemory object with the runs removed
+        # autoload=False to prevent recursively loading the memory
+        if metadata is not None and metadata != {}:
+            from griptape.memory.structure import BaseConversationMemory
+
+            return BaseConversationMemory.from_dict(
+                {
+                    **json.loads(thread_response.get("metadata")),
+                    "runs": [run.to_dict() for run in runs],
+                    "autoload": False,
```
Comment (collindutter) [inline griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py:118]: For visibility of offline conversation, why is this necessary in this Driver but not others?

File: tests/unit/drivers/memory/conversation/test_griptape_cloud_conversation_memory_driver.py [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+                    )
+                else:
+                    thread_id = args[0].split("/")[-1]
+                    return mocker.Mock(
+                        raise_for_status=lambda: None,
+                        json=lambda: {"metadata": TEST_CONVERSATION, "name": "test", "thread_id": "test_metadata"}
+                        if thread_id == "test_metadata"
+                        else {"name": "test", "thread_id": "test"},
+                    )
+            return None
```
Comment (collindutter) [inline tests/unit/drivers/memory/conversation/test_griptape_cloud_conversation_memory_driver.py:42]: Might be better to split this over two mocked requests clients.

File: tests/unit/drivers/memory/conversation/test_griptape_cloud_conversation_memory_driver.py [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+from griptape.memory.structure import BaseConversationMemory, ConversationMemory, Run, SummaryConversationMemory
+
+TEST_CONVERSATION = '{"type": "SummaryConversationMemory", "runs": [{"type": "Run", "id": "729ca6be5d79433d9762eb06dfd677e2", "input": {"type": "TextArtifact", "id": "1234", "value": "Hi There, Hello"}, "output": {"type": "TextArtifact", "id": "123", "value": "Hello! How can I assist you today?"}}], "max_runs": 2}'
+
+
+class TestGriptapeCloudConversationMemoryDriver:
+    @pytest.fixture(autouse=True)
+    def _mock_requests(self, mocker):
+        def get(*args, **kwargs):
+            if str(args[0]).startswith("https://cloud.griptape.ai/api/threads/"):
```
Comment (collindutter) [inline tests/unit/drivers/memory/conversation/test_griptape_cloud_conversation_memory_driver.py:16]: When would this ever not evaluate to `True` in the context of this test?

**Review by collindutter — CHANGES_REQUESTED** [review]

File: docs/griptape-framework/drivers/conversation-memory-drivers.md [hunk 23]
```diff
@@ -9,6 +9,14 @@ You can persist and load memory by using Conversation Memory Drivers. You can bu
 
 ## Conversation Memory Drivers
 
+### Griptape Cloud
+
+The [GriptapeCloudConversationMemoryDriver](../../reference/griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.md)  allows you to persist Conversation Memory in Griptape Cloud. It provides seamless integration with Griptape's cloud-based `Threads` and `Messages` resources.
```
[the remark this answers is not in the document]
Comment (vachillo) [reply docs/griptape-framework/drivers/conversation-memory-drivers.md:14]: where?

File: griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py [hunk 24]
```diff
@@ -0,0 +1,133 @@
...
+    )
+    api_key: Optional[str] = field(default=Factory(lambda: os.getenv("GT_CLOUD_API_KEY")))
+    headers: dict = field(
+        default=Factory(lambda self: {"Authorization": f"Bearer {self.api_key}"}, takes_self=True),
+        init=False,
+    )
+
+    def __attrs_post_init__(self) -> None:
+        if self.thread_id is None:
+            self.thread_id = os.getenv("GT_CLOUD_THREAD_ID", self._get_thread_id())
```
[the remark this answers is not in the document]
Comment (vachillo) [reply griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py:51]: better to do it here to ensure the api_key is loaded and validated first

File: griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py [hunk 25]
```diff
@@ -0,0 +1,133 @@
...
+
+    def load(self) -> Optional[BaseConversationMemory]:
+        # get the Messages from the Thread
+        messages_response = requests.get(self._get_url(f"/threads/{self.thread_id}/messages"), headers=self.headers)
+        messages_response.raise_for_status()
+        messages_response = messages_response.json()
+        if messages_response is None:
+            raise RuntimeError(f"Error getting messages for thread {self.thread_id}")
+
+        messages = sorted(messages_response.get("messages", []), key=lambda m: m["index"])
```
[the remark this answers is not in the document]
Comment (vachillo) [reply griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py:89]: this is a sanity check

File: griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py (hunk 16 above)
Comment (vachillo) [reply griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py:122]: should it be? its in a code branch and might not be needed

File: griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py (hunk 17 above)
Comment (vachillo) [reply griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py:112]: same here, its branched

File: griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py (hunk 16 above)
Comment (collindutter) [reply griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py:122]: You could extract to a separate function, but imports should always happen at the top of scope.

File: griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py (hunk 18 above)
Comment (vachillo) [reply griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py:117]: they need to be created either way, for the situation where the instance gets deserialized or if we are creating a new instance

File: griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (collindutter) [reply griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py:89]: Can we remove it then?

File: griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py (hunk 19 above)
Comment (vachillo) [reply griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py:70]: more flexible. it could be used as `metadata="test"` for all we care

File: griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py (hunk 20 above)
Comment (vachillo) [reply griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py:118]: no idea

File: griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py (hunk 17 above)
Comment (collindutter) [reply griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py:112]: Same solution should be applied. Move import to top of _a_ scope. Whether that's this one or a new function.

File: griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py (hunk 20 above)
Comment (vachillo) [reply griptape/drivers/memory/conversation/griptape_cloud_conversation_memory_driver.py:108]: updated other drivers with this fix

**Review by collindutter — APPROVED** [review]
